src/utils/listener/block_downloader.go

```diff
@@ -226,3 +198,124 @@ func (self *BlockDownloader) run() error {
 
 	return nil
 }
+
+func (self *BlockDownloader) downloadBlocks(height uint64, lastProcessedBlockHash arweave.Base64String) (out []*arweave.Block, err error) {
+	var mtx sync.Mutex
+	var wg sync.WaitGroup
+
+	peers := append(self.client.GetCachedPeers(), self.Config.Arweave.NodeUrl)
+	out = make([]*arweave.Block, 0, len(peers))
+	wg.Add(len(peers))
+
+	// Download blocks from all peers in parallel
+	for _, peer := range peers {
+		peer := peer
+		self.SubmitToWorker(func() {
+			block, err := self.downloadOneBlock(height, lastProcessedBlockHash, peer)
+			if err != nil {
+				block = nil
+			}
+			mtx.Lock()
+			if block != nil {
+				out = append(out, block)
+			}
+			mtx.Unlock()
+			wg.Done()
+		})
+	}
+
+	// Wait for all workers to finish
+	wg.Wait()
+
+	return
+}
+
+func (self *BlockDownloader) downloadOneBlock(height uint64, lastProcessedBlockHash arweave.Base64String, peer string) (block *arweave.Block, err error) {
+	ctx := context.WithValue(self.Ctx, arweave.ContextDisablePeers, true)
+	if len(peer) > 0 {
+		// Force using this peer if set
+		// Otherwise use arweave.net
+		ctx = context.WithValue(ctx, arweave.ContextForcePeer, peer)
+	}
+
+	block, resp, err := self.client.GetBlockByHeight(ctx, int64(height))
+	if err != nil {
+		self.Log.
+			WithError(err).
+			WithField("height", height).
+			WithField("peer", peer).
+			Error("Failed to download block")
+		return
+	}
+
+	if !block.IsValid() {
+		self.Log.
+			WithField("height", height).
+			WithField("peer", peer).
+			WithField("age", resp.Header().Get("Age")).
+			WithField("x-trace", resp.Header().Get("X-Trace")).
+			Error("Block hash isn't valid")
+		self.monitor.GetReport().BlockDownloader.Errors.BlockValidationErrors.Inc()
+		err = errors.New("block isn't valid")
+		return
+	}
+
+	if len(lastProcessedBlockHash) > 0 &&
+		!bytes.Equal(lastProcessedBlockHash, block.PreviousBlock) {
+		self.Log.
+			WithField("height", height).
+			WithField("age", resp.Header().Get("Age")).
+			WithField("x-trace", resp.Header().Get("X-Trace")).
+			WithField("last_processed_block_hash", lastProcessedBlockHash.Base64()).
+			WithField("previous_block", block.PreviousBlock.Base64()).
+			WithField("peer", peer).
+			Warn("Previous block hash isn't valid")
+		// This doesn't mean it's a bad block, we may have a fork
+	}
+
+	return
+}
+
+func (self *BlockDownloader) vote(blocks []*arweave.Block) (out *arweave.Block, err error) {
+	minNumberOfVotes := math.Min((math.Round(2.0 * float64(len(blocks)) / 3.0)), float64(len(blocks)))
```

Comment:
Not sure I follow this formula, but should we simply always require to have at least 3 same hashed from 5 different peers? or sth similar...

Reply:
this is the condition that there must be 2/3 votes, with 5 peers it means 4
this `math.Min` is redundant because 2/3*blocks is always smaller than blocks